### ISSUE TITLE
Passthru all arguments, including refresh token

### ISSUE THREAD

**Issue by evanhdc** [oauth-xx/oauth2#528]
Hi,

My API provider uses a standardized OAuth 2.0 token-based authentication flow. Now I get stuck on finding a way to refresh the access token.
```ruby
2.7.1 :001 > require 'oauth2'
2.7.1 :002 > client = OAuth2::Client.new('client_id', '', grant_type: 'refresh_token', site: 'site', token_url: 'token_url', redirect_uri: 'redirect_uri')
2.7.1 :003 > client.auth_code.authorize_url(access_type: 'offline')
2.7.1 :004 > tokens = client.auth_code.get_token('auth_code', 'grant_type': 'authorization_code')
2.7.1 :005 > tokens.refresh_token
 => nil
```
According to the API guide, the response attributes are
```json
{
  "access_token": "string",
  "expires_in": 0,
  "refresh_token": "string",
  "refresh_token_expires_in": 0,
}
```
But as shown in line 5, the `tokens.refresh_token` was `nil`. It looks like it's intentional per https://github.com/oauth-xx/oauth2/issues/321. How am I supposed to renew the access token, since a refresh token is required when using refresh token grant?

Here is the error. I waited until the access token expired, then tried to renew it:
```ruby
2.7.1 :006 > tokens.expired?
 => true 
2.7.1 :007 > new_tokens = tokens.refresh!
Traceback (most recent call last):
        1: from (irb):95
RuntimeError (A refresh_token is not available)
```

Did I miss anything?

**Comment by pboling** [comment]
I honestly do not know how to accomplish the flow you want, and have never done it myself.  If you do figure it out please document the solution here, or contribute documentation to the readme, or wiki.

This gem is on life support!

Any effort you could put into figuring out how to solve #321 would be greatly appreciated.  I doubt I will have time to invest in solving this bug myself.  This gem is too large a project to be maintained by a tiny team of people who only use it for narrow use cases.  We are always looking for more maintainers #307 !  Please lend a hand if you can.  I have a plan to dedicate more of my time to supporting this gem soon, but it keeps getting delayed, because the world is on fire, and I only have 24 hours in a day.

I volunteered to maintain this gem because it is important to the community, and the Ruby community matters to me, but I and the other maintainers are stretched thin.  I hope more people will step up to improve this tool!

**Comment by evanhdc** [comment]
I feel you and thank you for your contribution to the community!

Unfortunately I couldn't find a way and had to roll my own solution instead of using a gem.

There are two issues here:
First of all, which is also the issue I encountered above. The gem drops refresh token on the Authorization Grant response. The graph below is from [the Introduction section in RFC6749](https://tools.ietf.org/html/rfc6749#section-1.5)

```
  +--------+                                           +---------------+
  |        |--(A)------- Authorization Grant --------->|               |
  |        |                                           |               |
  |        |<-(B)----------- Access Token -------------|               |
  |        |               & Refresh Token             |               |
  |        |                                           |               |
  |        |                            +----------+   |               |
  |        |--(C)---- Access Token ---->|          |   |               |
  |        |                            |          |   |               |
  |        |<-(D)- Protected Resource --| Resource |   | Authorization |
  | Client |                            |  Server  |   |     Server    |
  |        |--(E)---- Access Token ---->|          |   |               |
  |        |                            |          |   |               |
  |        |<-(F)- Invalid Token Error -|          |   |               |
  |        |                            +----------+   |               |
  |        |                                           |               |
  |        |--(G)----------- Refresh Token ----------->|               |
  |        |                                           |               |
  |        |<-(H)----------- Access Token -------------|               |
  +--------+           & Optional Refresh Token        +---------------+

               Figure 2: Refreshing an Expired Access Token

   The flow illustrated in Figure 2 includes the following steps:

   (A)  The client requests an access token by authenticating with the
        authorization server and presenting an authorization grant.

   (B)  The authorization server authenticates the client and validates
        the authorization grant, and if valid, issues an access token
        and a refresh token.

  Omitting steps (C) to (F) as they are irrelevant...

   (G)  The client requests a new access token by authenticating with
        the authorization server and presenting the refresh token.  The
        client authentication requirements are based on the client type
        and on the authorization server policies.

   (H)  The authorization server authenticates the client and validates
        the refresh token, and if valid, issues a new access token (and,
        optionally, a new refresh token).
```
As we can see from step (B), the first refresh token must be included in the response so it can be used to renew access token in step (G), while the gem dropped it on purpose.

Secondly, as of the Optional Refresh Token in step (G), which is being discussed in #321. According to [the refresh token section in RFC6749](https://tools.ietf.org/html/rfc6749#section-1.5)
```
1.5.  Refresh Token

   Refresh tokens are credentials used to obtain access tokens.  Refresh
   tokens are issued to the client by the authorization server and are
   used to obtain a new access token when the current access token
   becomes invalid or expires, or to obtain additional access tokens
   with identical or narrower scope (access tokens may have a shorter
   lifetime and fewer permissions than authorized by the resource
   owner).  Issuing a refresh token is optional at the discretion of the
   authorization server.  If the authorization server issues a refresh
   token, it is included when issuing an access token
```
We can see that `Issuing a refresh token is optional at the discretion of the authorization server`. So if the authorization server decided to issue a refresh token, does it really make sense for the gem to drop it?

**Update**, I took another look at the section which is mentioned by @matt-glover in the same github issue, and now I understand why he decided to drop the refresh token.

He was referring to `4.4.3. Access Token Response` in the `Client Credentials Grant` section, which is a different workflow.  

If you would take a look at the [4.1. Authorization Code Grant section](https://tools.ietf.org/html/rfc6749#section-4.1.4), which clearly states that a refresh token could be included in the response.
```
4.1.4.  Access Token Response

   If the access token request is valid and authorized, the
   authorization server issues an access token and optional refresh
   token as described in Section 5.1.  If the request client
   authentication failed or is invalid, the authorization server returns
   an error response as described in Section 5.2.

   An example successful response:

     HTTP/1.1 200 OK
     Content-Type: application/json;charset=UTF-8
     Cache-Control: no-store
     Pragma: no-cache

     {
       "access_token":"2YotnFZFEjr1zCsicMWpAA",
       "token_type":"example",
       "expires_in":3600,
       "refresh_token":"tGzv3JOkF0XG5Qx2TlKWIA",
       "example_parameter":"example_value"
     }
```

In summary, it's "at the discretion of the authorization server" to return a refresh token or not. The gem should present any data received from the server to the user unopinionated. If a refresh token is not necessary, the server simply will not return it in the first place. I don't think the gem should make the decision to drop it or not.

**Comment by pboling** [comment]
@evanhdc Thanks for the work you put in here.  I will work on these changes for a version 2.1 release, now that v2 is finally (after 5 years!) done.

**Comment by nov** [comment]
Since #321 is mentioned here, I'm assuming you are using client credentials grant.
If so, you can get new access token anytime using client credentials.
You don't need to **refresh** them at all.

That's why RFC 6749 doesn't define refresh token in such flow.

**Comment by pboling** [comment]
@nov That makes sense, but an argument that was made elsewhere is that the server should know what it is doing.  If there is a refresh token this gem should send it along without attempting to enforce some arbitrary rule that only applies to a specific use case.

**Comment by nov** [comment]
yeah, passing given values without modification is OK as a library.
But the answer to the question "How to renew the access token, when oauth2 drops the refresh token?" is DO NOT REFRESH.
